### DISCUSSION TITLE
선착순 쿠폰 당첨자를 선정할 때, 응답 속도를 어떻게 빠르게 할 수 있을까?

### DISCUSSION THREAD

**Discussion by kmularise** [dailyge/dailyge-server#117] [Discussion]
## 📝 연관 PR

<br/><br/>

## ❓ 작성 이유
선착순 쿠폰 발급에 있어 발생할 수 있는 문제점들을 사전에 조사하고, 이를 해결하기 위한 여러가지 방안을 비교 분석한 후 프로젝트에 가장 적합한 방안을 선택하기 위해 작성하게 되었습니다.

<br/><br/>

## 1. 쿠폰 발급에 있어 발생할 수 있는 문제점
* 선착순 이벤트 관련 쿠폰과 상시 이벤트 관련 쿠폰을 구분할 필요가 있습니다.
* 쿠폰 과발급 현상 : 여러 사용자가 동시에 쿠폰을 요청할 때, 서버가 동일한 쿠폰을 여러번 발급하게 되는 경우입니다. 이는 전체 수량을 초과하여 쿠폰이 발급되는 상황을 초래할 수 있습니다.
* 쿠폰 미발급 현상 : 쿠폰이 남아 있음에도 불구하고 시스템이 일부 사용자에게 쿠폰을 발급하지 못하는 경우입니다. 이는 동시성 처리에서 경쟁 상태에 있을 때 발생할 수 있습니다.
이번에는 선착순 쿠폰 이벤트 처리에 초점을 맞춰서 해결방안을 떠올려보고 아키텍처를 설계해보았습니다.

<br>
<br>
<br>
<br>

### 아키텍처
![distributive](https://github.com/user-attachments/assets/4d784e34-7490-46e6-888c-01700c9738b4)
* 락 획득 : 사용자가 쿠폰 발급 신청 시 Redis에서 분산 락 요청
* 쿠폰 발급 : 락을 획득하면 쿠폰 발급 진행. 남아있는 쿠폰이 있는 경우 정상적으로 쿠폰 발급
* 락 해제 : 쿠폰 발급 작업이 완료되면 Redis에서 락 해제, 또는 타임아웃이 발생 시 락 해제

<br/><br/>

### 장점 및 단점
* 장점 : Redis의 스핀락과 비교했을 때 락을 획득하지 못한 경우 무한정 대기하는 것을 방지할 수 있습니다. 분산 시스템에서 여러 노드간의 동기화를 가능하게 합니다. 이로 인해 데이터의 안정성과 일관성을 유지할 수 있습니다.
* 단점 : 락을 획득하고 해제하는 과정에서 성능저하가 발생할 수 있습니다.

<br>
<br>
<br>
<br>

### 아키텍처

![pub_sub drawio](https://github.com/user-attachments/assets/2993ef09-ca9d-4785-b4e5-a9eb5dde1b23)

* publisher(발행자) : 쿠폰 발급 요청이 발생하면 Redis 채널로 쿠폰 발급 정보 관련 메시지 발행
* subscriber(구독자) : Redis 채널 구독, 메시지 수신 시 이를 처리하여 쿠폰 발급
* 쿠폰 데이터 저장소 : 쿠폰 수량을 실시간으로 관리
### 장점 및 단점
* 장점 : 비동기 처리를 통해 대규모 트래픽에서 요청을 빠르게 처리할 수 있습니다.
* 단점 : 네트워크 문제로 메시지 손실 가능성이 있고 메시지 처리 순서가 보장되지 않습니다.

<br>
<br>
<br>
<br>

### 아키텍처
![message_queue](https://github.com/user-attachments/assets/a156cdbe-cb5f-47dd-a1ba-49615e1a6146)
* RabbitMQ : 쿠폰 발급 시 메시지를 큐에 넣고, 순차적으로 처리
* Producer(생산자): RabbitMQ 큐에 매시지 발행
* Consumer(소비자): RabbitMQ에서 메시지 소비하여 쿠폰 발급을 처리
* 쿠폰 데이터 저장소: 쿠폰 수량을 관리
### 장점 및 단점
* 장점 : 메시지 전달 및 재시도가 관리되기에 메시지 손실을 방지할 수 있습니다.
* 단점 : 별도의 메시지 브로커를 관리해야 해서 추가적인 인프라가 필요합니다.

**Comment by devjun10** [comment]
## 📝 벌크 업데이트

오후에 구두로 회의한 내용입니다. 사용자 요청이 초당 2,500개씩 들어올 때, 이를 매번 레디스에 저장하면 2,500개의 요청이 전달됩니다. 물론 레디스 [공식 문서](https://redis.io/docs/latest/operate/oss_and_stack/management/optimization/benchmarks/)에 초당 10만 개의 요청까지 처리가 가능하다고 돼 있지만, 현재 저희는 **`설치형`** 으로 저 스펙의 서버를 레디스에 사용하고 있습니다. 따라서 가급적 효율적으로 사용자 요청을 처리하는 게 좋은데요, 이때, 사용자 요청을 **`애플리케이션 내부 큐`** 에 저장했다 **`벌크`** 로 레디스에 저장하는 방법을 고려할 수 있습니다.

![image](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FJhcZs%2FbtsJj9yqhZN%2FhMq1AkFecfy8pipmjfaIck%2Fimg.png)

<br/><br/><br/><br/>

다음과 같이요. 즉, 사용자 요청이 왔을 때, 애플리케이션 내부 큐에 해당 값들을 저장한 후, 이 값을 일정 주기로 벌크로 저장하는 것이죠. 이를 통해 **`레디스의 부하 감소`** 와 **`응답 속도`** 를 높일 수 있습니다. 예를 들어, 2,500개의 요청을 1,000개씩 벌크로 저장하면 레디스를 3번만 호출하게 됩니다. 매번, 2,500번 레디스를 호출하는 것보다 레디스의 부하가 줄어들겠죠? 또한, 애플리케이션 내부에 데이터를 저장하기 때문에 **`네트워크 호출`** 이 없어 빠르게 응답할 수 있고요. 운영체제 캐시를 생각해보면, CPU에 붙어있는 1차 캐시가 속도가 가장 빠른데, 이를 여기에 적용하면 애플리케이션과 바로 접해있는 인메모리 큐가 가장 빠를 테고요.

![image](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FD7dGt%2FbtsJj3ZklAc%2FD1FSK1wzU9kgNAugUfkCF1%2Fimg.png)

> 비슷한 기능은 [AWS SNS/SQS](https://docs.aws.amazon.com/ko_kr/sns/latest/dg/sns-batch-api-actions.html)도 지원하는데, 한 번 읽어보면 좋을 것 같습니다. 여기는 저장한 후, 한 번에 뽑고, 저희는 한 번에 저장하고 한 번에 뽑고. DB-to-DB 통신으로. 원리만 이해하면 될 것 같습니다.

<br/><br/><br/><br/>

아래는 부하 테스트를 했을 때, **`응답 속도`** 입니다. 일정 주기는 우리 애플리케이션의 **`응답 속도`** 를 보고 결정해야 할 것 같은데, 마지막 회원 가입/로그인 부하 테스트에서는 응답 속도가 평균적으로 **`0.35~0.4초`** 정도 나왔습니다. TPS는 약 1,800 정도였고요. RDB를 두어 번 호출한 것을 감안하면, 이런 과정이 없는, 이벤트를 레디스에 저장하는 요청은 더 빠르게 처리할 수 있을 것 같네요. 현재는 0.x초 ~ 1초 정도로 넉넉하게 해도 되지 않을까 하는데, 이건 다음 회의 때, 같이 한 번 이야기해 봐요.

![image](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2Fcsnjvz%2FbtsJiHwtriH%2FeYKls96pwj6jkFFYtagoh0%2Fimg.png)

> 즉, 한 서버에서 초당 몇 개의 요청을 얼마 만에 처리했는지를 감안해, 벌크 개수와 주기를 정하면 될 것 같습니다.

<br/><br/><br/><br/>

## ❗고려할 점

물론 고려할 점도 있는데, 이벤트 유실입니다. 애플리케이션 내부에 값을 저장하면 프로세스가 다운되는 순간 모든 데이터가 유실될 수 있기 때문이죠. 이는 [Graceful Shutdown](https://docs.spring.io/spring-boot/reference/web/graceful-shutdown.html) 으로 어느정도 해결 가능할 것 같은데, 마찬가지로 조금 더 고민해봐요. 

> 예를 들어, 1초 주기로 사용자 요청을 벌크로 저장한다고 하면, Graceful Shutdown 설정을 5초만 주더라도 모든 데이터를 레디스에 저장할 수 있습니다. 이론상은요. 😗

**Reply by kmularise** [reply]
한번 요청이 들어올 때마다 Redis에 요청을 보내게 되면, Redis 서버에서 처리해야 되는 연결이 많아지고 네트워크 호출로 인한 지연이 생길 수 있을 거 같습니다. 이를 인메모리큐를 사용하면 Redis 호출 수를 줄일 수 있어서 좋은 방안인 것 같습니다. 이때, 인메모리큐는 여러 스레드들에 의해 공유되어야 하기 때문에 힙 영역에 위치하고 동기화를 지원하는 큐 자료구조를 활용하는 것이 좋을 것 같습니다. 

Graceful Shutdown을 이용하면 기존에 처리 중인 요청이 완료될 수 있도록 허용하는 동시에 tocmat기준으로 새로운 요청이 네트워크 단에서 차단되기 때문에 애플리케이션 사용자가 쿠폰 발급 차단 시, 바로 작업이 실패했음을 알 수 있을 것 같습니다. 

인메모리 큐에서 Redis로 요청을 보내는 주기는 말씀하신 대로 개발 후 테스트를 해보면서 조정해봐야 할 것 같습니다. 인메모리 큐에서 Redis로 요청을 보내는 주기를 큐에 쌓여있는 데이터 개수도 고려하는 것을 생각해보았는데, 이렇게 되면 큐에 쌓여 있는 개수를 매번 확인하는 로직이 추가되어야 해서 이부분은 좀 더 생각해봐야할 거 같습니다.

Kafka나 RabbitMQ는 추가적인 인프라 구성이 필요하고 쿠폰 발급 정보에 대한 데이터를 영구적으로 저장할 필요성이 아직 없는 것 같습니다. Redis Pub/Sub는 메시지 유실 가능성이 있어서 데이터 일관성을 잘 보장할 수 있는 Redis 분산락 방식을 이용하는 방식으로 진행할 것 같습니다. 

말씀하신 내용 반영해서 진행해보도록 하겠습니다!

**Reply by devjun10** [reply]
분산락은 어디에 사용하나요? 오전에 이야기 된 내용이라면, 아래 회의 내용 정리 부탁드려요.

**Reply by kmularise** [reply]
## 1. 이전 상황
제가 원래 생각한 상황으로는 쿠폰 수량에 대해서 실시간으로 데이터 일관성이 지켜져야한다고 생각해서 각 발급 요청 벌크를 redis에 호출할 때 분산락을 사용해야 한다고 생각했습니다. 
<br></br>
## 2. 회의 상황
### 2-1. 쿠폰 발급에 대한 실시간 처리
회의한 결과, 쿠폰 발급에 대해서 실시간으로 처리해줄 필요가 있을까 생각했을 때, 실시간으로 처리하지 않아도 된다고 결론이 나왔습니다. 그 이유는 사용자가 쿠폰 발급에 따라서 사용자의 일정 관리 등 다른 행위가 영향을 받지 않기 때문입니다. 오히려 실시간으로 처리하는 것이 사용자가 쿠폰 발급을 기다리는 동안 아무런 행위도 할 수 없기 때문에 사용자 경험을 저해하기 때문입니다. 
<br></br>
### 2-2. 쿠폰 발급 참여 유무에 대한 유효성 검증
추가로, 쿠폰 발급 신청 시, 쿠폰 발급 신청에 대한 API를 호출하기 전에, 이미 쿠폰 발급을 신청했는지 여부를 검증하는 API를 호출해야 한다는 결론을 내렸습니다. 그 이유는 클라이언트 단에서 쿠키를 이용해서 참여 여부를 담아둘 수 있지만, 사용자가 조작할 수 있기 때문입니다. 그래서 쿠폰 발급 신청 여부를 확인하는 API를 통해, 우선 쿠키에 있는 쿠폰 발급 신청 여부를 확인하고 쿠폰 발급 신청이 아직 안된것으로 확인해서 만약 쿠폰 발급이 되었다면 Redis 호출을 통해 해당 유저의 쿠폰 발급을 확인하고 쿠폰 발급 신청 여부를 확인해야 합니다.

<br></br>
## 3. 결론
### 3-1. 아키텍처
최종적으로 결론 지은 아키텍처는 다음과 같습니다.
<br></br> 
![image](https://github.com/user-attachments/assets/060664fe-591f-467a-a830-3ed0668b9940)
<br></br>
1. 쿠폰 발급 요청이 들어오면 로드밸런서에서 가리킨 서버의 인메모리 큐에 쿠폰 발급 신청 정보가 쌓입니다.
2. 일정 주기마다 쿠폰 발급 요청에 대한 정보들을 redis에 저장합니다. 이로 인해 Redis 호출을 줄일 수 있습니다. 
3. 쿠폰 발급 처리 프로세스를 스케줄러에 등록해둡니다. 쿠폰 발급 수량에 대한 처리는 주기적으로 배치 작업을 이용해서 쿠폰 발급 처리 프로세스에서 Redis 정보를 이용해 당첨자를 추려낸 후, RDB에 쿠폰 발급 정보를 저장해두고자 합니다.
4. 발급 가능한 쿠폰 수량이 0이 된다면 해당 쿠폰 발급 처리 프로세스를 스케줄러에서 제거합니다.
<br></br>
 ### 3-2. 쿠폰 발급 신청 여부 검증 API 설계
![2024_0829_coupon_validation](https://github.com/user-attachments/assets/ad5f11cf-be2e-4772-b3b7-cf0e80f17d46)

쿠폰 발급 신청 화면 로드 시, 쿠폰을 이미 신청했다면 신청 버튼을 막아둬야 하는 상황입니다.
1. 클라이언트에서 쿠폰 발급 여부를 요청하는 API를 호출합니다.
2. 쿠폰 발급 여부 관련 쿠키 값이 true면 클라이언트에 응답을 반환합니다. 그렇지 않으면 3.으로 넘어갑니다.
3. 인메모리 큐를 조회해서 인메모리큐에 사용자 아이디가 포함되어 있으면 true를 반환합니다. 그렇지 않으면 4.으로 넘어갑니다.
4. Redis를 호출해서 Redis에 사용자 아이디가 포함되어 있으면 true를 반환하고 그렇지 않으면 false를 반환합니다.

**Reply by devjun10** [reply]
수고하셨어요. 이벤트 중복 참여 유무에 대한 결론은 아까 설명드린 내용 잘 이해하신 것 같아서 간단하게 요약만 할게요.

1. 이벤트 참여 페이지에 들어간 순간, 중복 참여 여부를 알 수 있게 합니다. 이를 통해 이벤트 참여 버튼 활성화 여부를 변경합니다.
2. 페이지에서 중복 참여를 막아도, API로 호출로 참여를 할 수도 있는데, 이를 막을 방도가 필요합니다. 이는 구두로 설명한 내용대로 비즈니스 로직으로 중복 참여를 제거할 수 있습니다. 레디스 SortedSet과 PriorityQueue 조합으로요. 이 부분 아까 잘 이해하신 것 같아, 결론만 남겨놓겠습니다.

**Reply by kmularise** [reply]
깔끔한 정리 감사해요! 논의한 방향대로 개발 진행해보도록 하겠습니다!

**Reply by kmularise** [reply]
추후에 회의를 진행해서 정리한 내용입니다. 
1. 처음 생각한 방안으로는 하나의 키와 하나의 sorted set에 사용자들을 담아둔다면 괜찮지 않을까 생각했습니다. 그치만 이때에 문제점이 사람들이 몰려서 당첨자 1000명인 쿠폰에 10만명이 발급 신청을 한다면  하나의 자료구조에 너무 많은 데이터를 담아두는 것이 문제라 생각했습니다. 따라서 요청 벌크(예: 요청 1000개) 당 하나의 key와 sorted set을 지정하고 발급 신청한 사용자 정보를 담아두기로 결정했습니다. 또한 redis INCR을 통해 요청 벌크가 레디스를 호출할 때마다 1을 증가시키는 방식으로 유일한 요청 벌크 아이디를 지정하고자 합니다.
`key: event{id}, value: sortedset` 이때, INCR과 같은 명령어는 원자적 연산이 보장되고 Redis는 코어가 싱글 스레드이기 때문에 동시성 문제가 발생하지 않게 됩니다.

2. 이때 하나의 키와 하나의 sorted set을 지정하고 사용자 정보를 시간 순으로 정렬하고 sorted set의 크기를 당첨자 수(1000개)만큼 제한하는 방법도 생각해봤습니다. 그렇지만 이때에는 사용자의 쿠폰 발급 신청 여부를 확인할 수 없다는 문제점이 있었습니다.

3. 1.에서 당첨자 추출은 첫번째 벌크 요청 sorted set과 두번째 sorted set에서 당첨자 수만큼을 추려내는 방식으로 병합한 다음 마지막 벌크 요청 sorted set까지 고려하여 당첨자 수만을 추려내는 다른 sorted set을 사용하는 방식을 사용하기로 했습니다. 이렇게 하면 각각의 요청 벌크 sorted set을 통해 사용자의 쿠폰 발급 신청 여부를 알 수 있고 당첨자 수를 추려낸 sorted set을 통해서 당첨자에게 쿠폰 발급을 진행할 수 있습니다. 이때 쿠폰 발급은 RDB를 통해 미리 발급을 해둡니다.

4. 쿠폰 발급 신청 여부를 확인할 때, 여러 키의 sorted set value에 대해서 Redis 호출을 줄이는 방식은 추후에 생각해보고자 합니다. 쿠폰 발급 중복 참여는 set 자료구조를 통해서 동일한 요소가 두번 들어가지 않게끔 구현할 수 있습니다. 

5. 0.1 초 정도로 주기를 잡고 인메모리 큐에서 벌크 요청을 Redis로 폴링(polling)을 하는 방식을 취하면 스레드가 작업을 처리하고 스레드 풀로 돌아와서 한정된 스레드로 요청을 빠르게 처리할 수 있을 것이라 기대됩니다. 다만 실제 테스트를 통해 좀더 최적의 주기를 찾아봐야할 것 같습니다.

**Reply by devjun10** [reply]
저희 정도 규모면 굳이 0.1초로 하지 않더라도 1초, 1.5초, 2초 정도로도 충분하지 않을까요 ㅎㅎ?

**Reply by kmularise** [reply]
@devjun10님 말씀처럼 쿠키를 조작하고 다시 쿠폰 발급 참여를 요청하는 시간과 트래픽이 1초 당 10만명 정도가 몰린다고 가정하면 1초 정도로 하는 것이 괜찮을 것 같아요. 주기를 너무 짧게 하는 것도 CPU의 컨텍스트 스위칭이 빈번하게 발생할 수 있고, Redis를 자주 호출하여 네트워크 호출로 인한 과부하가 일어날 수 있을 것 같습니다.
